### ISSUE TITLE
nan and maps non supported by current implementation of scaling factor for binned non-uniform axes

### ISSUE THREAD

**Issue by LMSC-NTappy** [hyperspy/hyperspy#2739]
The new implementation of scaling factor for binned non-uniform #2728 axes created some issues. Eg in the gaussian implementation:

https://github.com/hyperspy/hyperspy/blob/d1bcd1c540c9a7489a5527f32b15b5680e12f55b/hyperspy/_components/gaussian.py#L165-L168

the evaluation of `value2index(centre)` at line 168 relies on the fact that `_estimate_gaussian_parameters ` at line 165 will output a non-NaN value for centre. 

```python
s = hs.signals.Signal1D(np.zeros(100))

nuax = FunctionalDataAxis(expression="x**2",x=s.axes_manager.signal_axes[0],name='NUAX')
snu = s.deepcopy()
snu.axes_manager.set_axis(nuax,index_in_axes_manager=0)

g = hs.model.components1D.Gaussian(A=1,centre=80,sigma=10)
g.estimate_parameters(snu,20.,900.)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-51-6b634f634055> in <module>
----> 1 g.estimate_parameters(snu,20.,900.)

c:\users\xxxxxxx\documents\git\hyperspy\hyperspy\_components\gaussian.py in estimate_parameters(self, signal, x1, x2, only_current)
    166                                                               only_current)
    167         scaling_factor = axis.scale if axis.is_uniform \
--> 168                          else np.gradient(axis.axis)[axis.value2index(centre)]
    169         if only_current is True:
    170             self.centre.value = centre

c:\users\xxxxxxx\documents\git\hyperspy\hyperspy\axes.py in value2index(self, value, rounding)
    567             return (np.abs(self.axis - value)).argmin()
    568         else:
--> 569             index = int(value)
    570             if self.size > index >= 0:
    571                 return index

ValueError: cannot convert float NaN to integer
```
This, regardless of whether the axis is binned or not

Additionnaly, this implementation also fails for signals with navigation dimensions > 0.  It is because `value2index(centre)` at line 168 relies on the fact that `_estimate_gaussian_parameters ` at line 165 will not output an array of values, which it will.

```python 
s = hs.signals.Signal1D(np.zeros((10,10,100)))

nuax = FunctionalDataAxis(expression="x**2",x=s.axes_manager.signal_axes[0],name='NUAX', bin)
snu = s.deepcopy()
snu.axes_manager.set_axis(nuax,index_in_axes_manager=2)

g = hs.model.components1D.Gaussian()

g.estimate_parameters(s,20.,40.)
True

g.estimate_parameters(snu,20.,400.)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-69-413522c6418a> in <module>
----> 1 g.estimate_parameters(snu,20.,40.)

c:\users\xxxxxx\documents\git\hyperspy\hyperspy\_components\gaussian.py in estimate_parameters(self, signal, x1, x2, only_current)
    166                                                               only_current)
    167         scaling_factor = axis.scale if axis.is_uniform \
--> 168                          else np.gradient(axis.axis)[axis.value2index(centre)]
    169         if only_current is True:
    170             self.centre.value = centre

c:\users\xxxxxx\documents\git\hyperspy\hyperspy\axes.py in value2index(self, value, rounding)
    564         if value is None:
    565             return None
--> 566         if self.low_value <= value <= self.high_value:
    567             return (np.abs(self.axis - value)).argmin()
    568         else:

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
Also regardless if the axes is binned or not.

I propose to fix that by:
-Ensuring that the computation of scaling_factor is performed only after checking that the axis is binned
-Ensuring that `centre` values are passed one-by-one to value2index
-Ensuring that  `NaN` values are forwarded by the function without disturbing its operation

 I will try to implement a patch asap

**Comment by ericpre** [comment]
Thanks @LMSC-NTappy, I noticed similar issues and started to fix at i, but haven't finished it!

> I propose to fix that by:
-Ensuring that the computation of scaling_factor is performed only after checking that the axis is binned
-Ensuring that centre values are passed one-by-one to value2index
-Ensuring that NaN values are forwarded by the function without disturbing its operation
>
> I will try to implement a patch asap

Your suggestion sounds very good. Regarding `value2index`, passing array should be supported for `BaseDataAxis`; this is one of the error mentioned above and this could be fixed in separate PR.
Please go ahead with a PR, alternatively, you could review a PR fixing these issues.

This is very useful that you are using this `non_uniform_axes`, because we need this branch to get some exposure - some edges are still fairly rough but it is not far from being ready!

**Comment by ericpre** [comment]
This is how far, I have been on this issue: https://github.com/ericpre/hyperspy/tree/nua_estimate_parameters_fixes - the test I added to cover this issue is still failing.

**Comment by LMSC-NTappy** [comment]
> This is how far, I have been on this issue: https://github.com/ericpre/hyperspy/tree/nua_estimate_parameters_fixes - the test I added to cover this issue is still failing.

Ok I'll work from this then. Thanks!

**Comment by LMSC-NTappy** [comment]
I close this issue because the pull request succesfully merged
